datafusion/core/src/physical_plan/hash_join.rs

```diff
@@ -791,6 +826,109 @@ fn build_join_indexes(
     }
 }
 
+fn apply_join_filter(
+    left: &RecordBatch,
+    right: &RecordBatch,
+    join_type: JoinType,
+    left_indices: UInt64Array,
+    right_indices: UInt32Array,
+    filter: &JoinFilter,
+) -> Result<(UInt64Array, UInt32Array)> {
+    if left_indices.is_empty() {
+        return Ok((left_indices, right_indices));
+    };
+
+    let (intermediate_batch, _) = build_batch_from_indices(
+        filter.schema(),
+        left,
+        right,
+        PrimitiveArray::from(left_indices.data().clone()),
+        PrimitiveArray::from(right_indices.data().clone()),
+        filter.column_indices(),
+    )?;
+
+    match join_type {
+        JoinType::Inner | JoinType::Left => {
+            // For both INNER and LEFT joins, input arrays contains only indices for matched data.
+            // Due to this fact it's correct to simply apply filter to intermediate batch and return
+            // indices for left/right rows satisfying filter predicate
+            let filter_result = filter
+                .expression()
+                .evaluate(&intermediate_batch)?
+                .into_array(intermediate_batch.num_rows());
+            let mask = as_boolean_array(&filter_result);
+
+            let left_filtered = PrimitiveArray::<UInt64Type>::from(
+                compute::filter(&left_indices, mask)?.data().clone(),
```

Comment:
I think here, we apply filter after `build_batch_from_indices` means : apply filter after join operator. Is there a chance filter before join.
if wrong plz correct me.

Reply:
Yes, at physical level filter always applied after actual join operation.

And you're right - there are cases when it's fine to filter inputs before join step - but here it's responsibility of logical planner / optimizer - they both are able to check if predicate (or its part) could be pushed before join - closer to scan.

So, when it comes to physical join, logical plan supposed to contain only `t1.field < t2.field`-like predicates which could not be applied before join

Reply:
I agree with @korowa  that moving filters out of the `On` clause is better left to the planner and optimizer.

This is an important point though -- perhaps we could add a comment to the docstring of the HashJoin operator explaining that `filter` should ideally only contain expressions that can not be pushed to the inputs of the join, one way or the other

Reply:
@korowa @alamb ❤️ Thanks your explanations! I think after merge this Pr, we should add a check in `filter_push_down.rs` to improve this situation.